### PR TITLE
Feature/expire

### PR DESCRIPTION
**Pull request by hakod**
Hi @dr-dimitru,

I added an expiration feature for while code is still active. Wasn't sure whether to add timestamp for expiration date since the other PR already had that. Please let me know if I should change it to use a timestamp as well. Thanks again.

### CODE REVIEW

**Comment by dr-dimitru** [comment]
@hakod thank you for the quick PR.

 1. How is it going to work after page reload?
 2. Yes, don't mind another PRs
 2. Please let's discuss an idea before implementing it

**Comment by hakod** [comment]
@dr-dimitru No problem! Yeah, I was concerned that the setTimeout would be interrupted. Would a timestamp be the best route and should there be a setInterval to check for expiration every so often? Thanks for reviewing.

**Comment by dr-dimitru** [comment]
@hakod I believe we should utilize cookie's expiration in case if cookies are used, use records with `.___expireAt` prefix holding expirations timestamp, having loop via `setInterval(()=>{}, 256)` for LocalStorage, and use `setTimeout` for JS implementation when LocalStorage nor Cookies are available. Does it sounds clear?

**Comment by hakod** [comment]
@dr-dimitru Yes, I think I understand it.

File: client-storage.js [hunk 1]
```diff
@@ -216,21 +216,20 @@ ClientStorage.prototype.get = function (key) {
...
-        this.ls.removeItem(key)
-      }, time)
+      this.ls.setItem(key + "._expiredAt", expire);
     } else if (this.cookies) {
       this.cookies.set(key, this.__escape(value));
-      setTimeout(() => {
-        this.cookies.remove(key, null, window.location.host)
-      }, time)
+      let seconds = time / 1000;
+      document.cookie = key + "=" + value + "; max-age=" + seconds + "; path=/";
```
Comment (dr-dimitru) [inline client-storage.js]: Please use `this.cookies` [library](https://github.com/VeliovGroup/Client-Storage/blob/master/client-storage.js#L78) here. You can update `this.cookies.set` method to support expiration date

File: client-storage.js [hunk 2]
```diff
@@ -216,21 +216,20 @@ ClientStorage.prototype.get = function (key) {
...
     } else {
       this._data[key] = value;
       setTimeout(() => {
         delete this._data[key];
       }, time)
     }
+    return true
```
Comment (dr-dimitru) [inline client-storage.js:246]: I'd use `else if` here

File: client-storage.js [hunk 3]
```diff
@@ -242,6 +241,21 @@ ClientStorage.prototype.set = function (key, value, time) {
   return true;
 };
 
+function LSExpire() {
+  setInterval(() => {
+    for (var key in localStorage) {
+      if (key.slice(-11) === "._expiredAt") {
```
Comment (dr-dimitru) [inline client-storage.js]: I'd use ES6 `.includes('._expiredAt')` method and single quotes on the strings

**Comment by hakod** [comment]
@dr-dimitru Added the cookie expiration using max-age and the prefix to store expiration times. Let me know if there are any changes needed. Thanks.

File: client-storage.js [hunk 4]
```diff
@@ -242,6 +241,21 @@ ClientStorage.prototype.set = function (key, value, time) {
...
+        if (window.localStorage.getItem(key) < new Date().getTime()) {
+          window.localStorage.removeItem(key);
+          window.localStorage.removeItem(key.replace("._expiredAt", ""))
+        }
+      }
+    }
+  }, 2000)
+}
+
+LSExpire();
```
Comment (dr-dimitru) [inline client-storage.js]: Only if `localStorage` is used

Comment (dr-dimitru) [reply client-storage.js]: Should be in constructor

**Review by dr-dimitru — CHANGES_REQUESTED** [review]

**Comment by dr-dimitru** [comment]
@hakod thank you, added notes

**Comment by hakod** [comment]
@dr-dimitru Thanks for the feedback, I really appreciate it. The last note was a bit difficult for me to fix. Should setInterval be triggered whenever page loads? How could I get that working?

File: client-storage.js [hunk 5]
```diff
@@ -152,33 +152,33 @@ function ClientStorage(driver) {
...
-  case 'localStorage':
-    if (this.LSSupport) {
-      this.ls = window.localStorage || localStorage;
-    } else {
-      console.warn('ClientStorage is set to "localStorage", but it is not supported on this browser');
-    }
-    break;
-  case 'cookies':
-    if (this.cookies) {
+    case 'localStorage':
```
Comment (dr-dimitru) [inline client-storage.js]: Please, no indentation inside `switch(){}`

**Review by dr-dimitru — CHANGES_REQUESTED** [review]
Check inside constructor if `localStorage` is supported and `setInterval`

**Comment by hakod** [comment]
@dr-dimitru OK, I think I got it. Let me know, thanks.

**Comment by dr-dimitru** [comment]
Dear @hakod ,

Thank you for your PR and attempt, final solution released as [`v3.0.0`](https://github.com/VeliovGroup/Client-Storage/releases/tag/3.0.0).
Do not hesitate to share your opinion 👨‍💻

**Comment by hakod** [comment]
Hi @dr-dimitru, 
It is interesting you used a while statement instead of setInterval to remove localstorage. Wondering what the reason for that change was.

**Comment by dr-dimitru** [comment]
@hakod not using `while` nor `setInterval`. Timers are unreliable in Browser's runtime environment. Timers will also pollute JS runtime. So, I'm just checking for expiration date [here](https://github.com/VeliovGroup/Client-Storage/blob/master/client-storage.js#L243) and remove it from storage if expired, it's also happening when *ClientStorage* instance is initiated

**Comment by hakod** [comment]
OK I think I understand it now. Would it delete without a page refresh too?

**Comment by dr-dimitru** [comment]
No, only upon request if expired or on page init. Very similar to how cookies and expiration works in most of browsers

**Comment by hakod** [comment]
That makes a lot more sense. Thanks.

**Pull request closed without merging**